### PR TITLE
Handle private container registry while downloading blobs

### PR DESCRIPTION
**Pull request by adarsh-zededa**
Added registry to the image reference while downloading.
Signed-off-by: adarsh-zededa <adarsh@zededa.com>

### CODE REVIEW

File: pkg/pillar/zedUpload/ociutil/oci.go [hunk 1]
```diff
@@ -100,10 +100,7 @@ func PullBlob(registry, repo, hash, localFile, username, apiKey string, maxsize
 
 	// The OCI distribution spec only uses /blobs/ endpoint for layers or config, not index or manifest.
 	// I have no idea why you cannot get a manifest or index from the /blobs endpoint, but so be it.
-	image := repo
-	if hash != "" {
-		image = fmt.Sprintf("%s@%s", repo, hash)
-	}
+	image := fmt.Sprintf("%s/%s", registry, repo)
```
Comment (deitch) [inline pkg/pillar/zedUpload/ociutil/oci.go:103]: Will this work? We pass the actual hash in, and here we ignore it, leading us to get nothing for L119:

```go
		d, ok := ref.(name.Digest)
```

Comment (adarsh-zededa) [reply pkg/pillar/zedUpload/ociutil/oci.go:103]: As per our current flow, `repo` already has an image hash, so we don't need to add an image hash.

Comment (deitch) [reply pkg/pillar/zedUpload/ociutil/oci.go:103]: I thought that sometimes it does, and sometimes it doesn't? Yes, I know, I wrote much of that `zedUpload/ociutil/`, but doesn't mean I remember it. Is it not possible to sometimes have no hash? I don't mean that we never actually do that, but that it is possible to hit that code flow?

Comment (eriknordmark) [reply pkg/pillar/zedUpload/ociutil/oci.go:103]: @adarsh-zededa if that is the case, then why are you changing image on line 105? I'm confused.

Comment (adarsh-zededa) [reply pkg/pillar/zedUpload/ociutil/oci.go:103]: @nordmark I actually updated the PR, but for some reason it's not showing up here, I'm changing the `image` at line 103.
This is the last commit: https://github.com/lf-edge/eve/pull/1242/commits/f03e8b07e879ab6ab07740a913d66df6f5a5e949

Comment (eriknordmark) [reply pkg/pillar/zedUpload/ociutil/oci.go:103]: And you change the image variable on line 105 (old code) which doesn't seem to match your explanation

Comment (adarsh-zededa) [reply pkg/pillar/zedUpload/ociutil/oci.go:103]: I handled image at line 105. Thanks @eriknordmark!

**Comment by deitch** [comment]
Good to get our heads around the requirements @eriknordmark . 

`PullBlob()` is supposed to get anything that is provided, with or without a hash. It is the equivalent of `docker pull <ref>`. But it also supports explicitly passing a hash, whether it is provided in the `<ref>` or not. So it should be able to handle all of the standard use cases:

* `nginx@sha256:1234`
* `nginx:latest@sha256:1234`
* `nginx:latest` 

Plus it needs to be able to handle passing the hash separately.

Here are your use cases:

> repo = "nginx@sha256:1234" + hash = ""

As you said, "the code will do the right thing."

> repo = "nginx@sha256:1234" + hash = "sha256:1234"

It should handle it, although it probably should be cross-referencing them to ensure they are the same.

> repo = "nginx:latest" + hash = "sha256:1234"
> repo = "nginx" + hash = "sha256:1234"

Yes. These two are the same. As soon as you use a hash, the tag is ignored.

> repo = anything + hash = "sha256:" (i.e. invalid hash)

Yes, but it should reject it. And it does, if we append the hash, since this will return an error:

```go
	ref, err := name.ParseReference(image)
	if err != nil {
		return 0, fmt.Errorf("parsing reference %q: %v", image, err)
	}
```

**Comment by eriknordmark** [comment]
> > repo = "nginx:latest" + hash = "sha256:1234"
> > repo = "nginx" + hash = "sha256:1234"
> 
> Yes. These two are the same. As soon as you use a hash, the tag is ignored.
> 

I don't know what the underlying package supports; but I suspect need to remove any tag when we add the sha at the end

**Comment by deitch** [comment]
> I don't know what the underlying package supports; but I suspect need to remove any tag when we add the sha at the end

We shouldn't need to, as this is what parses it all:

```go
	ref, err := name.ParseReference(image)
	if err != nil {
		return 0, fmt.Errorf("parsing reference %q: %v", image, err)
	}
```

And should do the right thing.

I whipped up a little program to test it:

```go
package main

import (
	"fmt"

	"github.com/google/go-containerregistry/pkg/name"
)


func main() {
	names := [][]string{
		{"nginx:abc", "valid tag, no hash"},
		{"nginx:abc@sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d", "valid tag, valid hash"},
		{"nginx:abc@sha256:abcd", "valid tag, invalid hash"},
		{"nginx:abc@sha256:", "valid tag, start of hash"},
		{"nginx", "no tag, no hash"},
		{"nginx@sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d", "no tag, valid hash"},
		{"nginx@sha256:abcd", "no tag, invalid hash"},
		{"nginx@sha256:", "no tag, start of hash"},
	}
	for _, image := range names {
		ref, err := name.ParseReference(image[0])
		fmt.Printf("'%s' '%s' %#v '%v'\n", image[0], image[1], ref, err)
	}
}
```

And the output:

```
'nginx:abc' 'valid tag, no hash' name.Tag{Repository:name.Repository{Registry:name.Registry{insecure:false, registry:""}, repository:"nginx"}, tag:"abc", original:"nginx:abc"} '<nil>'
'nginx:abc@sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d' 'valid tag, valid hash' name.Digest{Repository:name.Repository{Registry:name.Registry{insecure:false, registry:"index.docker.io"}, repository:"library/nginx"}, digest:"sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d", original:"nginx:abc@sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d"} '<nil>'
'nginx:abc@sha256:abcd' 'valid tag, invalid hash' <nil> 'could not parse reference: nginx:abc@sha256:abcd'
'nginx:abc@sha256:' 'valid tag, start of hash' <nil> 'could not parse reference: nginx:abc@sha256:'
'nginx' 'no tag, no hash' name.Tag{Repository:name.Repository{Registry:name.Registry{insecure:false, registry:""}, repository:"nginx"}, tag:"", original:"nginx"} '<nil>'
'nginx@sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d' 'no tag, valid hash' name.Digest{Repository:name.Repository{Registry:name.Registry{insecure:false, registry:"index.docker.io"}, repository:"library/nginx"}, digest:"sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d", original:"nginx@sha256:c7e7c316f879aa68e98165cfe8252e5b2c3540f4a527949eb1eb34e96afbd46d"} '<nil>'
'nginx@sha256:abcd' 'no tag, invalid hash' <nil> 'could not parse reference: nginx@sha256:abcd'
'nginx@sha256:' 'no tag, start of hash' <nil> 'could not parse reference: nginx@sha256:'
```

**Pull request merged**